File: app/src/main/java/com/extensionbox/app/modules/FapCounterModule.java
package com.extensionbox.app.modules;

import android.app.NotificationManager;
import android.content.Context;

import androidx.core.app.NotificationCompat;

import com.extensionbox.app.Prefs;
import com.extensionbox.app.R;
import com.extensionbox.app.SystemAccess;

import java.util.Calendar;
import java.util.LinkedHashMap;
import java.util.Locale;

public class FapCounterModule implements Module {

    private Context ctx;
    private boolean running = false;

    @Override public String key() { return "fap"; }
    @Override public String name() { return "Fap Counter"; }
    @Override public String emoji() { return "🍆"; }
    @Override public String description() { return "Self-monitoring counter & streak tracker"; }
    @Override public boolean defaultEnabled() { return false; }
    @Override public boolean alive() { return running; }
    @Override public int priority() { return 100; }

    @Override
    public int tickIntervalMs() {
        return ctx != null ? Prefs.getInt(ctx, "fap_interval", 60000) : 60000;
    }

    @Override
    public void start(Context c, SystemAccess sys) {
        ctx = c;
        running = true;
        checkDayRollover();
    }

    @Override
    public void stop() {
        running = false;
    }

    @Override
    public void tick() {
        checkDayRollover();
    }

    /**
     * Called externally (e.g., from notification action or dashboard) to increment counter.
     */
    public void increment() {
        if (ctx == null) return;
        int today = Prefs.getInt(ctx, "fap_today", 0);
        today++;
        Prefs.setInt(ctx, "fap_today", today);

        // Update monthly total
        int monthly = Prefs.getInt(ctx, "fap_monthly", 0);
        Prefs.setInt(ctx, "fap_monthly", monthly + 1);

        // Update all-time total
        int allTime = Prefs.getInt(ctx, "fap_all_time", 0);
        Prefs.setInt(ctx, "fap_all_time", allTime + 1);

        // Reset streak (days without)
        Prefs.setInt(ctx, "fap_streak", 0);
        Prefs.setInt(ctx, "fap_last_day", getDayOfYear());
    }

    /**
     * Check if day changed and update streak/daily counter accordingly.
     */
    private void checkDayRollover() {
        if (ctx == null) return;
        int currentDay = getDayOfYear();
        int lastDay = Prefs.getInt(ctx, "fap_last_check_day", -1);

        if (lastDay == -1) {
            // First run
            Prefs.setInt(ctx, "fap_last_check_day", currentDay);
            return;
        }

        if (currentDay != lastDay) {
            // Day changed
            int todayCount = Prefs.getInt(ctx, "fap_today", 0);

            if (todayCount == 0) {
                // No faps yesterday → increase streak
                int streak = Prefs.getInt(ctx, "fap_streak", 0);
                Prefs.setInt(ctx, "fap_streak", streak + 1);
            }

            // Save yesterday's count
            Prefs.setInt(ctx, "fap_yesterday", todayCount);

            // Reset daily counter
            Prefs.setInt(ctx, "fap_today", 0);

            // Check month rollover
            int currentMonth = Calendar.getInstance().get(Calendar.MONTH);
            int lastMonth = Prefs.getInt(ctx, "fap_last_month", -1);
            if (lastMonth != -1 && currentMonth != lastMonth) {
                Prefs.setInt(ctx, "fap_prev_monthly", Prefs.getInt(ctx, "fap_monthly", 0));
                Prefs.setInt(ctx, "fap_monthly", 0);
            }
            Prefs.setInt(ctx, "fap_last_month", currentMonth);

            Prefs.setInt(ctx, "fap_last_check_day", currentDay);
        }
    }

    public int getTodayCount() {
        return ctx != null ? Prefs.getInt(ctx, "fap_today", 0) : 0;
    }

    private int getDayOfYear() {
        return Calendar.getInstance().get(Calendar.DAY_OF_YEAR);
    }

    @Override
    public String compact() {
        int today = ctx != null ? Prefs.getInt(ctx, "fap_today", 0) : 0;
        int streak = ctx != null ? Prefs.getInt(ctx, "fap_streak", 0) : 0;
        if (streak > 0) return "🍆" + today + " 🔥" + streak + "d";
        return "🍆" + today + " today";
    }

    @Override
    public String detail() {
        if (ctx == null) return "🍆 No data";
        int today = Prefs.getInt(ctx, "fap_today", 0);
        int streak = Prefs.getInt(ctx, "fap_streak", 0);
        int monthly = Prefs.getInt(ctx, "fap_monthly", 0);

        StringBuilder sb = new StringBuilder();
        sb.append(String.format(Locale.US, "🍆 Today: %d", today));
        if (streak > 0) sb.append(String.format(Locale.US, " • 🔥 Streak: %dd clean", streak));
        sb.append(String.format(Locale.US, "\n   Monthly: %d", monthly));
        return sb.toString();
    }

    @Override
    public LinkedHashMap<String, String> dataPoints() {
        LinkedHashMap<String, String> d = new LinkedHashMap<>();
        if (ctx == null) return d;

        int today = Prefs.getInt(ctx, "fap_today", 0);
        int yesterday = Prefs.getInt(ctx, "fap_yesterday", 0);
        int streak = Prefs.getInt(ctx, "fap_streak", 0);
        int monthly = Prefs.getInt(ctx, "fap_monthly", 0);
        int allTime = Prefs.getInt(ctx, "fap_all_time", 0);

        d.put("fap.today", String.valueOf(today));
        d.put("fap.yesterday", String.valueOf(yesterday));
        d.put("fap.streak", streak > 0 ? streak + " days 🔥" : "0");
        d.put("fap.monthly", String.valueOf(monthly));
        d.put("fap.all_time", String.valueOf(allTime));
        return d;
    }

    @Override
    public void checkAlerts(Context ctx) {
        int dailyLimit = Prefs.getInt(ctx, "fap_daily_limit", 0);
        if (dailyLimit <= 0) return;

        int today = Prefs.getInt(ctx, "fap_today", 0);
        boolean alertFired = Prefs.getBool(ctx, "fap_limit_fired", false);

        if (today >= dailyLimit && !alertFired) {
            fireAlert(ctx, 2010,
                    "🍆 Daily Limit Reached",
                    "You've reached your daily limit of " + dailyLimit + ". Take a break!");
            Prefs.setBool(ctx, "fap_limit_fired", true);
        }

        // Reset alert flag at day change
        if (today == 0 && alertFired) {
            Prefs.setBool(ctx, "fap_limit_fired", false);
        }
    }

    private void fireAlert(Context c, int id, String title, String body) {
        try {
            NotificationManager nm = (NotificationManager)
                    c.getSystemService(Context.NOTIFICATION_SERVICE);
            NotificationCompat.Builder b = new NotificationCompat.Builder(c, "ebox_alerts")
                    .setSmallIcon(R.drawable.ic_notif)
                    .setContentTitle(title)
                    .setContentText(body)
                    .setPriority(NotificationCompat.PRIORITY_HIGH)
                    .setAutoCancel(true);
            nm.notify(id, b.build());
        } catch (Exception ignored) {}
    }
}

File: app/src/main/java/com/extensionbox/app/modules/FapCounterModule.kt
package com.extensionbox.app.modules

import android.app.NotificationManager
import android.content.Context
import androidx.core.app.NotificationCompat
import com.extensionbox.app.Prefs
import com.extensionbox.app.R
import com.extensionbox.app.SystemAccess
import java.util.Calendar
import java.util.LinkedHashMap
import java.util.Locale

class FapCounterModule : Module {

    private var ctx: Context? = null
    private var running = false

    override fun key(): String = "fap"
    override fun name(): String = "Fap Counter"
    override fun emoji(): String = "🍆"
    override fun description(): String = "Self-monitoring counter & streak tracker"
    override fun defaultEnabled(): Boolean = false
    override fun alive(): Boolean = running
    override fun priority(): Int = 100

    override fun tickIntervalMs(): Int = ctx?.let { Prefs.getInt(it, "fap_interval", 60000) } ?: 60000

    override fun start(ctx: Context, sys: SystemAccess) {
        this.ctx = ctx
        running = true
        checkDayRollover()
    }

    override fun stop() {
        running = false
    }

    override fun tick() {
        checkDayRollover()
    }

    fun increment() {
        val c = ctx ?: return
        val today = Prefs.getInt(c, "fap_today", 0) + 1
        Prefs.setInt(c, "fap_today", today)

        val monthly = Prefs.getInt(c, "fap_monthly", 0) + 1
        Prefs.setInt(c, "fap_monthly", monthly)

        val allTime = Prefs.getInt(c, "fap_all_time", 0) + 1
        Prefs.setInt(c, "fap_all_time", allTime)

        Prefs.setInt(c, "fap_streak", 0)
        Prefs.setInt(c, "fap_last_day", getDayOfYear())
    }

    private fun checkDayRollover() {
        val c = ctx ?: return
        val currentDay = getDayOfYear()
        val lastDay = Prefs.getInt(c, "fap_last_check_day", -1)

        if (lastDay == -1) {
            Prefs.setInt(c, "fap_last_check_day", currentDay)
            return
        }

        if (currentDay != lastDay) {
            val todayCount = Prefs.getInt(c, "fap_today", 0)
            if (todayCount == 0) {
                val streak = Prefs.getInt(c, "fap_streak", 0)
                Prefs.setInt(c, "fap_streak", streak + 1)
            }
            Prefs.setInt(c, "fap_yesterday", todayCount)
            Prefs.setInt(c, "fap_today", 0)

            val currentMonth = Calendar.getInstance().get(Calendar.MONTH)
            val lastMonth = Prefs.getInt(c, "fap_last_month", -1)
            if (lastMonth != -1 && currentMonth != lastMonth) {
                Prefs.setInt(c, "fap_prev_monthly", Prefs.getInt(c, "fap_monthly", 0))
                Prefs.setInt(c, "fap_monthly", 0)
            }
            Prefs.setInt(c, "fap_last_month", currentMonth)
            Prefs.setInt(c, "fap_last_check_day", currentDay)
        }
    }

    fun getTodayCount(): Int = ctx?.let { Prefs.getInt(it, "fap_today", 0) } ?: 0

    private fun getDayOfYear(): Int = Calendar.getInstance().get(Calendar.DAY_OF_YEAR)

    override fun compact(): String {
        val today = ctx?.let { Prefs.getInt(it, "fap_today", 0) } ?: 0
        val streak = ctx?.let { Prefs.getInt(it, "fap_streak", 0) } ?: 0
        return if (streak > 0) "🍆$today 🔥${streak}d" else "🍆$today today"
    }

    override fun detail(): String {
        val c = ctx ?: return "🍆 No data"
        val today = Prefs.getInt(c, "fap_today", 0)
        val streak = Prefs.getInt(c, "fap_streak", 0)
        val monthly = Prefs.getInt(c, "fap_monthly", 0)

        val sb = StringBuilder()
        sb.append("🍆 Today: $today")
        if (streak > 0) sb.append(" • 🔥 Streak: ${streak}d clean")
        sb.append("\n   Monthly: $monthly")
        return sb.toString()
    }

    override fun dataPoints(): LinkedHashMap<String, String> {
        val d = LinkedHashMap<String, String>()
        val c = ctx ?: return d

        d["fap.today"] = Prefs.getInt(c, "fap_today", 0).toString()
        d["fap.yesterday"] = Prefs.getInt(c, "fap_yesterday", 0).toString()
        val streak = Prefs.getInt(c, "fap_streak", 0)
        d["fap.streak"] = if (streak > 0) "$streak days 🔥" else "0"
        d["fap.monthly"] = Prefs.getInt(c, "fap_monthly", 0).toString()
        d["fap.all_time"] = Prefs.getInt(c, "fap_all_time", 0).toString()
        return d
    }

    override fun checkAlerts(ctx: Context) {
        val dailyLimit = Prefs.getInt(ctx, "fap_daily_limit", 0)
        if (dailyLimit <= 0) return

        val today = Prefs.getInt(ctx, "fap_today", 0)
        val alertFired = Prefs.getBool(ctx, "fap_limit_fired", false)

        if (today >= dailyLimit && !alertFired) {
            fireAlert(ctx, 2010, "🍆 Daily Limit Reached", "You've reached your daily limit of $dailyLimit. Take a break!")
            Prefs.setBool(ctx, "fap_limit_fired", true)
        }

        if (today == 0 && alertFired) {
            Prefs.setBool(ctx, "fap_limit_fired", false)
        }
    }

    private fun fireAlert(c: Context, id: Int, title: String, body: String) {
        try {
            val nm = c.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            nm.notify(id, NotificationCompat.Builder(c, "ebox_alerts")
                .setSmallIcon(R.drawable.ic_notif)
                .setContentTitle(title)
                .setContentText(body)
                .setPriority(NotificationCompat.PRIORITY_HIGH)
                .setAutoCancel(true).build())
        } catch (ignored: Exception) {
        }
    }
}
